File: frontend/src/api/memo.ts
import apiClient from './client'

export interface Memo {
  id: string
  user_id: string
  content: string
  summary: string
  tags: string | null
  created_at: string
  updated_at: string
}

export interface MemoCreateRequest {
  content: string
  summary?: string
  tags?: string[]
}

export interface MemoUpdateRequest {
  content?: string
  summary?: string
  tags?: string[]
}

export interface MemoListResponse {
  memos: Memo[]
  total: number
}

export const memoApi = {
  // 创建备忘录
  async create(data: MemoCreateRequest): Promise<Memo> {
    const response = await apiClient.post('/memos', data)
    return response.data
  },

  // 获取备忘录列表
  async list(tags?: string): Promise<MemoListResponse> {
    const params = tags ? { tags } : {}
    const response = await apiClient.get('/memos', { params })
    return response.data
  },

  // 获取单个备忘录
  async get(id: string): Promise<Memo> {
    const response = await apiClient.get(`/memos/${id}`)
    return response.data
  },

  // 更新备忘录
  async update(id: string, data: MemoUpdateRequest): Promise<Memo> {
    const response = await apiClient.put(`/memos/${id}`, data)
    return response.data
  },

  // 删除备忘录
  async delete(id: string): Promise<void> {
    await apiClient.delete(`/memos/${id}`)
  },

  // 解析标签字符串为数组
  parseTags(tags: string | null): string[] {
    if (!tags) return []
    return tags.split(',').map(tag => tag.trim()).filter(tag => tag.length > 0)
  },

  // 格式化标签数组为字符串
  formatTags(tags: string[]): string {
    return tags.join(', ')
  }
}

File: backend/app/schemas/memo.py
"""备忘录相关的Pydantic schemas"""
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime


class MemoCreateRequest(BaseModel):
    """创建备忘录请求"""
    content: str = Field(..., min_length=1, description="备忘录内容")
    summary: Optional[str] = Field(default=None, description="内容摘要")
    tags: Optional[List[str]] = Field(default=None, description="标签列表")
    
    class Config:
        json_schema_extra = {
            "example": {
                "content": "今天学习了Python的异步编程，理解了async/await的工作原理",
                "summary": "学习Python异步编程",
                "tags": ["学习", "Python"]
            }
        }


class MemoResponse(BaseModel):
    """备忘录响应"""
    id: str = Field(..., description="备忘录ID")
    user_id: str = Field(..., description="用户ID")
    content: str = Field(..., description="备忘录内容")
    summary: str = Field(..., description="内容摘要")
    tags: Optional[str] = Field(default=None, description="标签（逗号分隔）")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")
    
    class Config:
        from_attributes = True
    
    @property
    def tags_list(self) -> List[str]:
        """将标签字符串转换为列表"""
        if not self.tags:
            return []
        return [tag.strip() for tag in self.tags.split(',') if tag.strip()]


class MemoUpdateRequest(BaseModel):
    """更新备忘录请求"""
    content: Optional[str] = Field(default=None, min_length=1, description="备忘录内容")
    summary: Optional[str] = Field(default=None, description="内容摘要")
    tags: Optional[List[str]] = Field(default=None, description="标签列表")


class MemoListResponse(BaseModel):
    """备忘录列表响应"""
    memos: list[MemoResponse] = Field(..., description="备忘录列表")
    total: int = Field(..., description="总数")
